### ISSUE TITLE
Have samplers store more data about simulation history

### ISSUE THREAD

**Issue by jchodera** [choderalab/perses#194]
We should have the samplers store more information about the sampler history in a NetCDF file or something.

Ideas so far:
* current chemical state key
* proposed chemical state key
* log acceptance probability
* accept/reject statistics
* current `Topology` 
* current configuration and box volume
* log weights
* reduced potential energy

What other data should we keep track of?

**Comment by pgrinaway** [comment]
potential energies?

**Comment by jchodera** [comment]
Good idea! Let me update the issue with a list.

**Comment by pgrinaway** [comment]
I will update this in a bit, but sort of spitballing here:

```python

class PersesOutput(object):

    def __init__(self, output_filename):
        self._filename = output_filename

    def write_configuration(self, positions, topology, reduced_potential=None, thermodynamic_state=None):
        pass

    def write_chemical_state(self, topology, chemical_state_key, reduced_potential=None, thermodynamic_state=None):
        pass

    def write_SAMS_weights(self, weights, target_weights, sams_sampler_name):
        pass

    def write_metadata(self, data, writer_name):
        pass

    def increment_iteration(self, sampler_name):
        pass

    def write_logp_accept(self, logp_accept, old_chemical_state, new_chemcial_state, accepted):
        pass

    def get_current_iteration(self, sampler_name):
        pass
```

I don't know if there should be anything more specific. I had imagined that `GeometryEngine` et al. could write out their specific debug info via the `write_metadata` method. The object would keep track of the current iterations, so that everything would be organized.

**Comment by pgrinaway** [comment]
Also, I should add that I would change the base class constructors to also take a reference to a `PersesOutput` object, which would give each class an opportunity to insert whatever metadata it needs to.

**Comment by steven-albanese** [comment]
+1 on the output class, this looks great

**Comment by jchodera** [comment]
Can we generalize the API a bit? We may want to cram a ton of debug info in there, including logP for various substeps, configurations for various substeps or parts of substeps, timings of substeps, and other random information. If we need to extend the API for every little statistic we might need for debugging, it will hurt more than it helps to add this layer of abstraction.

What if we focus on a few general API commands to store some common types of per-iteration information, and have the API help us by creating the NetCDF variables under the hood if they don't already exist? Then we just need to standardize on the naming of various objects in the file that are important for analysis. 

We can also make use of object groups, which are like directories. Each module or Perses can use a separate group (directory) to keep its namespace isolated from others.

**Comment by jchodera** [comment]
Do we also want to be able to store System objects, or are we confident we can reconstruct those exactly from the Topology object, even without storing Forcefield object?

**Comment by jchodera** [comment]
For example, we could have something like
```python
# Store a configuration, or a configuration in a sequence of configurations (e.g. for NCMC)
# This should probably automagically store the topology object so we can reconstruct PDB files or trajectories.
# If iteration is None, the object is a singleton.
storage.store_configuration(group, name, topology, positions, iteration=None, frame_index=0, nframes=1)
# Store a serializable python object.
# If iteration is None, the object is a singleton.
storage.store_object(group, name, object, iteration=None)
# Store some numpy array in native form.
storage.store_array(group, name, array, iteration=None, dtype=None)
```

**Comment by pgrinaway** [comment]
>Can we generalize the API a bit? We may want to cram a ton of debug info in there, including logP for various substeps, configurations for various substeps or parts of substeps, timings of substeps, and other random information. If we need to extend the API for every little statistic we might need for debugging, it will hurt more than it helps to add this layer of abstraction.

Absolutely, that was just a stab.

>What if we focus on a few general API commands to store some common types of per-iteration information

How about:

* Configurations

* Thermodynamic states (including system/topology?)

* Log probabilities

* Metadata (other debug and timing information)

>We can also make use of object groups, which are like directories. Each module or Perses can use a separate group (directory) to keep its namespace isolated from others.

That's what I was thinking too. Sorry--left this out.


>Do we also want to be able to store System objects,

Yep, forgot about this.

To your example:

One of the things I was considering was how to keep track of iteration index--there are several different counters for various modules. I was thinking the `PersesOutput` instance could actually keep track of these, so that data would automatically be assigned to the correct indices. Otherwise, how would, for instance, `GeometryEngine` know what the current iteration of the `ExpandedEnsembleSampler` or `SAMSSampler` is?

Also, what is the purpose of the `store_object` and `store_array` functions? Wouldn't they just be thin wrappers around the NetCDF library? Or do we want them there for precisely that--edge cases where we want to dump extra information?

**Comment by jchodera** [comment]
As we discussed yesterday, I think the goal is:
* Each module of the code should store its data in a separate NetCDF group
* Use a simple API with only a few general classes of variables: configurations (likely with corresponding `Topology` so we can write PDB files), numpy arrays, serialized python objects; the module passes the group and variable names
* When a new variable that doesn't exist is written through an API call, the variable and corresponding dimensions should be created under the hood if they don't already exist---this will make it easy to instrument the code without having to worry about adding all the initialization stuff
* Use a simple iteration management scheme, though it may be best for each module to manage its own iterations since otherwise it will get very confusing if we have 10 MCMC iterations per expanded ensemble iteration; 3 expanded ensemble iterations per SAMS iteration, etc.

I think something like the API proposed is all we need:
```
class NetCDFStorage(object):

    def __init__(self, filename, mode='a'):
        self._filename = filename
        # Open NetCDF file here...

    def write_configuration(self, modname, varname, positions, topology, iteration=None, frame=None, nframes=None):
        """Write a configuration (or one of a sequence of configurations) to be stored as a native NetCDF array

        Parameters
        ---
        modname : str
            The name of the module in the code writing the variable
        varname : str
            The variable name to be stored
        positions : simtk.unit.Quantity of size [natoms,3] with units compatible with angstroms
            The positions to be written
        topology : simtk.openmm.Topology object
            The corresponding Topology object
        iteration : int, optional, default=None
            The local iteration for the module, or `None` if this is a singleton
        frame : int, optional, default=None
            If these coordinates are part of multiple frames in a sequence, the frame number
        nframes : int, optional, default=None
            If these coordinates are part of multiple frames in a sequence, the total number of frames in the sequence

        """
        pass

    def write_object(self, modname, varname, object, iteration=None):
        """Serialize a Python object

        Parameters
        ---
        modname : str
            The name of the module in the code writing the variable
        varname : str
            The variable name to be stored
        object : object
            The object to be serialized
        iteration : int, optional, default=None
            The local iteration for the module, or `None` if this is a singleton
        """
        pass

    def write_array(self, modname, varname, array, iteration=None):
        """Write a numpy array as a native NetCDF array

        Parameters
        ---
        modname : str
            The name of the module in the code writing the variable
        varname : str
            The variable name to be stored
        array : numpy.array of arbitrary dimension
            The numpy array to be written
        iteration : int, optional, default=None
            The local iteration for the module, or `None` if this is a singleton
        """
        pass
```
What do you think?  Would that make it easy to store what we need?

**Comment by pgrinaway** [comment]
That looks good! I like this idea. Thanks!

**Comment by jchodera** [comment]
Implemented in #220.
